README
vecscreen_plus_taxonomy
Author: Alejandro Schaffer
Documentation help and code review: Eric Nawrocki 

github: https://github.com/aaschaffer/vecscreen_plus_taxonomy.git
Version: 0.17
December 2020
------------------
README

This file describes

-- a script called from_vecscreen_to_summary.pl and its constituent
programs

and
-- a summary classification script compare_vector_matches_wtaxa.pl
that can be used to run vecscreen and parse its outputs, including
taxonomy information.

There is a related repository 

github: https://github.com/aaschaffer/generate_vecscreen_candidates.git

Outline of this file:

BACKGROUND
SETTING UP ENVIRONMENT VARIABLES
SAMPLE RUN, USAGE AND OUTPUT OF from_vecscreen_to_summary.pl
SAMPLE RUN, USAGE AND OUTPUT OF compare_vector_matches_w_taxa.pl
METHODS
TEST SCRIPT
RELEVANT FILES
CREATING THE TAXONOMY FILE
AN ADDITIONAL TAXONOMY HELPER PROGRAM
COMPILING vecscreen AND srcchk ON LINUX, FROM SCRATCH

**************
**BACKGROUND**
**************

vecscreen is the established NCBI program to identify matches between
(query) sequences and (subject) vectors in UniVec. These matches may
represent (true) vector contamination, but experience has shown that
there can be many false positives.  A primary reason for false
positives is that the query and the matching subject segment come from
the same genus or closely related genera. Therefore, knowing the genus
of the query sequence is helpful to interpret the output.  For what
follows below, it is relevant that vecscreen distinguishes matches by
two characteristics:

1) Location: Internal or Terminal 
A match is Terminal if and only if it includes a nucleotide within 25
positions of either end of the query; otherwise, the match is Internal.

2) Strength: Strong, Moderate, or Weak
A match is Strong if either: it is Terminal with a raw score of at
least 24 or it is Internal with a raw score of at least 30.

A match is Moderate if either: it is Terminal with a raw score in the
interval [19,23] or Internal with a raw score in the interval [25,29].

A match is Weak if it is Terminal with a raw score in the interval
[16,18] or Internal with a raw score in the interval [23,24].

vecscreen also reports Internal alignments with raw scores in the
range [16, 22] when there is also a reportable match for the same
(query, vector) pair. The score range [16,22] is below the Weak range
match for Internal matches. In the script
from_vecscreen_to_summary.pl, these below-Weak Internal matches are
assigned the level None.

More information about vecscreen can be found at
https://www.ncbi.nlm.nih.gov/tools/vecscreen/about/
In this document, 'vecscreen' refers to the command-line version, not
the Web page version.

The programs and data in this vecscreen_plus_taxonomy repository are
for prospective analysis of nucleotide sequences, especially
sequences that one might want to submit to GenBank.

The related repository 

github: https://github.com/aaschaffer/generate_vecscreen_candidates.git

does a preprocessing step relevant to only retrospective analysis of
sequences already in a database set up for BLAST, such as the non-redundant
(nr) database in GenBank. In retrospective analysis, one wants to start
by searching the database for sequences that may be expected to have vecscreen
matches, so as to avoid running vecscreen on the entire database. 

************************************
**SETTING UP ENVIRONMENT VARIABLES**
************************************

Before you can run from_vecscreen_to_summary.pl or
compare_vector_matches_wtaxa.pl you will need to update some of your
environment variables. To do this, add the following four lines to
your .bashrc file (if you use bash shell) or .cshrc file (if you use C
shell or tcsh). The .bashrc or .cshrc file will be in your home
directory. To determine what shell you use type 'echo $SHELL', if it
returns '/bin/bash', then update your .bashrc file, if it returns
'/bin/csh' or '/bin/tcsh' then update your .cshrc file.

The 4 lines to add to your .bashrc file, ***but make sure that you
replace PATH/TO/VEC/PLUS with the directory path to the directory
created when you cloned the vecscreen_plus_taxonomy github repository.

-----------
export VECPLUSDIR="PATH/TO/VEC/PLUS"
export PERL5LIB="$VECPLUSDIR:$PERL5LIB"
export PATH="$VECPLUSDIR/scripts:$PATH"
export BLASTDB="$VECPLUSDIR/univec-files:$BLASTDB"
-----------

The 4 lines to add to your .cshrc file:
-----------
setenv VECPLUSDIR "PATH/TO/VEC/PLUS"
setenv PERL5LIB "$VECPLUSDIR":"$PERL5LIB"
setenv PATH "$VECPLUSDIR/scripts":"$PATH"
setenv BLASTDB "$VECPLUSDIR/univec-files":"$BLASTDB"
-----------

Then, after adding those 4 lines, execute this command:
source ~/.bashrc
OR
source ~/.cshrc

If PERL5LIB was not already defined (you will know, if you get an error
message when you run the above 'source' command): 
use instead
export PERL5LIB="$VECPLUSDIR"
for .bashrc, OR
setenv PERL5LIB "$VECPLUSDIR"
for .cshrc.
at line 2 out of 4. 

Similarly, if BLASTDB was not already defined (you will know, if you
get an error message when you run the above 'source' command): 
use instead
export BLASTDB="$VECPLUSDIR/univec-files"
for .bashrc, OR
setenv BLASTDB "$VECPLUSDIR/univec-files"
for .cshrc.
at line 4 out of 4. 

To check that your environment variables are properly set up do the
following four commands:
echo $VECPLUSDIR
echo $PERL5LIB
echo $PATH
echo $BLASTDB

The first command should return only:
PATH/TO/VEC/PLUS

The second echo command should return a potentially longer
string that begins with the same path:
PATH/TO/VEC/PLUS

The third echo command should return a (potentially longer) string
that begins with:
PATH/TO/VEC/PLUS/scripts

The fourth echo command should return a (potentially longer) string
that begins with:
PATH/TO/VEC/PLUS/univec-files

If that is not the case, please email Alejandro Schaffer
(alejandro.schaffer@nih.gov). If you do see the expected output, the
following sample run should work.

Finally, you will need to gunzip two exectuable files in
PATH/TO/VEC/PLUS/scripts, and make the files executable (in case they
are not already executable). Perform the following commands:

> cd $VECPLUSDIR/scripts
> gunzip vecscreen.gz
> gunzip srcchk.gz
> chmod +x vecscreen
> chmod +x srcchk

After this, you should be able to successfully complete the sample
runs below, as well as the test script in the TEST SCRIPT section
below. 

****************************************************************
**SAMPLE RUN, USAGE AND OUTPUT OF from_vecscreen_to_summary.pl**
****************************************************************

Here is an example command that will run from_vecscreen_to_summary.pl
on the fasta file myseqs.fa using the taxonomy file included in 
PATH/TO/VEC/PLUS/info-files/taxonomy_tree_wlevels.txt. 

> from_vecscreen_to_summary.pl --output_root mytest --input_fasta $VECPLUSDIR/test-files/test.input_sequence_file.fa --input_taxa $VECPLUSDIR/info-files/taxonomy_tree_wlevels.txt --combine_output --verbose > mytest.out

This command is contained in the file
$VECPLUSDIR/scripts/sample_run.sh. In fact, it is better to just run
that shell script file, because it will also check that the expected
output is correct:

> sh $VECPLUSDIR/scripts/sample_run.sh

When you run the above 'sh' command, you should see output like this:
--
comparing expected output test-files/expected.output_combined_wtaxonomy.txt to observed output mytest.output_combined_wtaxonomy.txt
SUCCESS: Files are identical
--

After running the above from_vecscreen_to_summary.pl command, the
one-line per sequence output is in the file
mytest.output_combined_wtaxonomy.txt. This 11-column format is
described below under 'Verbose output format (enabled with --verbose
option): 11 columns'.

The output of the script to stdout, which describes briefly what the
script is doing, is in the file mytest.out.

The input sequence file
($VECPLUSDIR/test-files/test.input_sequence_file.fa) can be any
nucleotide sequence file in FASTA format.

The taxonomy file ($VECPLUSDIR/info-files/taxonomy_tree_wlevels.txt)
is a file in a special format that includes taxonomy information based
on NCBI's taxonomy. This can either be created by the user from the
NCBI taxonomy file, or the user can use the provided file
taxonomy_tree_wlevels.txt, which was created in April 2018. If you are
using this in 2018, the provided file above should be fine. After
that, you should create a new up-to-date file. Instructions for doing
that are provided below in the section CREATING THE TAXONOMY FILE.

The options that can be provided to from_vecscreen_to_summary.pl are:
  --input_fasta <s> : REQUIRED: file name <s> with sequences in fasta format 
  --input_taxa <s>  : REQUIRED: file name <s> mapping vecscreen matches to taxa
  --output_root <s> : REQUIRED: output files will be named starting with <s>
  --verbose         : output 11 columns instead of 5
  --combine_output  : combine internal and terminal matches
  --keep	    : keep all intermediate files (e.g. vecscreen output)

The --input_fasta <s> and --input_taxa <s> and --output_root <s>
options are required when running from_vecscreen_to_summary.pl, while
--verbose and --combine_output and --keep are optional.  By default
--verbose and --combine_output and --keep are all turned off.

The --combine_output option determines whether one or two files of
output are produced.  Currently the names of the output files are
partly fixed at:
<output_root>.output_combined_wtaxonomy.txt (if --combine_output is applied) 
<output_root>.output_internal_wtaxonomy.txt (if --combine_output is not applied) 
<output_root>.output_terminal_wtaxonomy.txt (if --combine_output is not applied)

The option --verbose determines how many columns of output are
produced, as described in the OUTPUT section below. It does not cause
more diagnostic output to be printed during the execution of the
script. The extra columns are important for classifying some matches
as true or false contamination.

The option --keep determines whether the output file from running
vecscreen within the script is kept (--keep used) or deleted (--keep
not used, default).

from_vecscreen_to_summary.pl will create one line of tabular output
per vecscreen hit in the input sequence file. There are two possible
output formats. For both formats, columns are separated by tabs.

--------------------------------
Default output format: 5 columns
--------------------------------
By default (if --verbose is not used) then the format of those lines
will be the following five columns:

Column 1: Accession of query
Column 2: Genus of query if known, or 1 otherwise
Column 3: Matching vector, starting with uv|
Column 4: One end of the alignment in the vector
Column 5: The other end of the alignment in the vector

This 5 column format was agreed on for internal NCBI usage in JIRA
ticket SM-187.

----------------------------------------------------------------
Verbose output format (enabled with --verbose option): 11 columns
----------------------------------------------------------------
If --verbose is used, then each line of output will include the
following 11 columns:

Column 1:  Accession of query
Column 2:  Genus of query if known, or 1 otherwise
Column 3:  Species of query if known, or 1 otherwise
Column 4:  Lower end of the alignment in the query
Column 5:  Upper end of the alignment in the query
Column 6:  Matching vector, starting with uv|
Column 7:  One end of the alignment in the vector
Column 8:  The other end of the alignment in the vector
Column 9:  The strength of this vecscreen match 
Column 10: The strength of the strongest vecscreen match for this query
Column 11: Whether there is any dangling part (called "Suspect" by
           vecscreen) at either end of the query

A dangling part is an unmatched segment of <= 25 nucleotides.

This alternative 11-column format has been shown to be useful for some
purposes, such as correcting vector-contaminated sequences in GenBank.
Another circumstance in which the 11-column format
is essential is if there is an input sequence that has a known species
(not 1 in column 3) but do not have a known genus (1 in column 2).

*********************************************************************
**SAMPLE RUN, USAGE, AND OUTPUT OF compare_vector_matches_w_taxa.pl**
*********************************************************************

The compare_vector_matches_w_taxa.pl script takes as input the
one-line per-sequence output file from
from_vecscreen_to_summary.pl. That input file must be the 11-column
output of from_vecscreen_to_summary.pl that is created when the
--verbose option is used for that script.

Here is the example usage of compare_vector_matches_wtaxa.pl:

compare_vector_matches_wtaxa.pl \ 
--input_summary $VECPLUSDIR/test-files/test.sample_input_final_step.txt \
--input_taxa $VECPLUSDIR/info-files/taxonomy_tree_wlevels.txt \
--input_artificial_vectors $VECPLUSDIR/info-files/artificial_whole_sequences.txt \
--input_artificial_segments $VECPLUSDIR/info-files/artificial_intervals.txt \
--input_univec_sources $VECPLUSDIR/info-files/biological_exclusions.txt \
--input_amr $VECPLUSDIR/info-files/UniVec10_vs_amr_distinct_intervals.txt \
--input_sequences $VECPLUSDIR/test-files/test.sample_candidates.fa \
--outfile my_output_final_step.txt

This command is contained in the file
$VECPLUSDIR/test-files/sample_compare_run.sh. As above, it's better to
just run that shell script file, because it will also check that the
expected output is correct:

> sh $VECPLUSDIR/scripts/sample_compare_run.sh

When you run the above 'sh' command, you should see output like this:
--
comparing expected output test-files/expected.output_final_step.txt to observed output my_outputfinal_step.txt
SUCCESS: Files are identical
--

After running this compare_vector_matches_wtaxa.pl command, the file
'my_output_final_step.txt' will include the output of
compare_vector_matches_wtaxa.pl which is the file
test-files/sample_input_final_step.txt with 3 additional columns:

Column 12: the classification of the match
Column 13: Most pertinent taxid of the vector interval 
Column 14: Lowest common ancestor of column 2 and column 13

The possible classifications in column 12 are currently:
NO_DATA, TRUE_ARTIFICIAL, TRUE_ARTIFICIAL_MICROSAT, FALSE_AMR,
FALSE_BIOLOGICAL, TRUE_BIOLOGICAL, TRUE_MICROSAT, LIKELY_FALSE,BACTERIAL.
These are explained below.
In practice, one mainly wants to distinguish between:
{TRUE_ARTIFICIAL, TRUE_ARTIFICIAL_MICROSAT, TRUE_BIOLOGICAL, TRUE_MICROSAT} which are true contamination,
versus
{FALSE_AMR, FALSE_BIOLOGICAL, LIKELY_FALSE,BACTERIAL}, which are false contamination.


In some cases, the classification TRUE_BIOLOGICAL may be too bold and
this can be seen because Column 14 is not much higher up the taxonomy
tree than column 13.  When this happens, the conservation of the
vector source needs to be propagated from genus_level_exclusions.txt
to one of the higher-level files*:
   superkingdom_level_exclusions.txt
   kingdom_level_exclusions.txt
   phylum_level_exclusions.txt
   class_level_exclusions.txt
   order_level_exclusions.txt
   family_level_exclusions.txt
   tribe_level_exclusions.txt

* Please email alejandro.schaffer@nih.gov if you find any examples of
vector sources that should be propagated to a higher level of taxonomy.

***********
**METHODS**
***********

Given the input taxonomy file and an input sequence file in FASTA
format, from_vecscreen_to_summary.pl will do the following:

1) Run vecscreen on the input FASTA-formatted sequence file to
   identify high-scoring matches to known vector sequences in UniVec
   in the input sequence file.

2) Parse the vecscreen output into two tab-delimited files for
   internal and terminal matches by calling parse_vecscreen.pl.

3) Optionally combine the two summary files into one by calling
   combine_summaries.pl.

4) Add taxonomy information to the vecscreen output by calling srcchk
   and add_taxonomy.pl.

compare_vector_matches_wtaxa.pl uses the six sets of data files listed
to classify each vecscreen match.  This program is separate because an
in-house usage needed somewhat different I/O specifications to fit
into an existing software framework.

***************
**TEST SCRIPT**
***************
There is a 'test' script included in the vecscreen_plus_taxonomy
distribution that you should run to make sure that everything is set
up correctly, in addition to doing the two example sample runs above.

To run the test script, execute the following command:
> $VECPLUSDIR/scripts/test_vecscreen_plus_taxonomy_scripts.pl

You should see the following output:
--
Checking that required input files exist ... done.
Testing combine_summaries.pl             ... done.
Testing add_taxonomy.pl                  ... done.
Testing from_vecscreen_to_summary.pl     ... done.
Testing compare_vector_matches_wtaxa.pl  ... done.
# All tests passed.
# SUCCESS
--

If you do not see this output, make sure that you've set up your
environment variable $VECPLUSDIR correctly as explained above in the
'SETTING UP ENVIRONMENT VARIABLES' section. If you still have
problems, email alejandro.schaffer@nih.gov.

******************
**RELEVANT FILES**
******************

Several executable files are required for from_vecscreen_to_summary.pl
and compare_vector_matches_wtaxa.pl to work. Two of these executables
were developed by others and must be downloaded and installed
separately outside of NCBI. All of these files are included in the
vecscreen_plus_taxonomy github repository, so you do not need to
create or move any files in order to get from_vecscreen_to_summary.pl
to work. After cloning the git repository, these files will be in the
PATH/TO/VEC/PLUS/scripts directory.

The first two files are NCBI executable programs that were not authored by
Alejandro Schaffer:

1. vecscreen and associated UniVec database
   Identifies vector contamination in input sequences.

   Within NCBI, vecscreen can be found at
   /netopt/ncbi_tools64/c++.stable/ReleaseMT/bin/vecscreen

   At NCBI, to add this directory to your path execute this command: 
   > ln -s /netopt/ncbi_tools64/c++.stable/ReleaseMT/bin/vecscreen .

   For users outside NCBI, we provide a gzipped executable of vecscreen
   for 64-bit Linux computers in
   scripts/vecscreen.gz
   Run
   gunzip vecscreen.gz
   chmod +x vecscreen

   and make sure that vecscreen is on the execution path (which it should
   be if you followed the steps above in SETTING UP ENVIRONMENT VARIABLES)

   vecscreen requires that the BLASTable database UniVec be
   accessible.  This means that these files must be in a directory
   that is contained in your $BLASTDB environment variable. If you
   followed the instructions in the SETTING UP ENVIRONMENT VARIABLES
   section above, you have already added the appropriate directory
   (PATH/TO/VEC/PLUS/univec-files) to $BLASTDB.  The UniVec database
   is represented (for purposes of vecscreen) in the three files:

   univec-files/UniVec.nhr
   univec-files/UniVec.nin
   univec-files/UniVec.nsq
   Do not try to edit the three UniVec files under any circumstances.

   See also the section COMPILING vecscreen AND srcchk ON LINUX, FROM SCRATCH

2. srcchk 
   Determines the taxonomy of input sequences, with respect the NCBI
   taxonomy tree.

   Within NCBI, srcchk can be found at
   /netopt/ncbi_tools64/bin/srcchk

   At NCBI, to add this directory to your path execute this command:
   > ln -s /netopt/ncbi_tools64/bin/srcchk .

   For users outside NCBI, we provide a gzipped executable of srcchk
   for 64-bit Linux computers in
   scripts/srcchk.gz
   Run
   gunzip srcchk.gz
   chmod +x srcchk

   and make sure that srcchk is on the execution path (which it should
   be if you followed the steps above in SETTING UP ENVIRONMENT VARIABLES)

   See also the section COMPILING vecscreen AND srcchk ON LINUX, FROM SCRATCH

The next 6 files were authored by Alejandro Schaffer:

3. from_vecscreen_to_summary.pl 
   This is the main script that coordinates the work by calling the
   other executables listed below.

4. parse_vecscreen.pl
   Auxiliary script called by from_vecscreen_to_summary.pl that parses
   the vecscreen output.
   
5. combine_summaries.pl
   Auxiliary script called by from_vecscreen_to_summary.pl that
   combines two different output formats of parse_vecscreen.pl

6. add_taxonomy.pl
   Auxiliary script called by from_vecscreen_to_summary.pl that adds
   taxonomy information to the output of parse_vecscreen.pl.

7. assign_levels_to_taxonomy.pl
   Independent script that is used to create a taxonomy file from the
   NCBI taxonomy. The file produced by assign_levels_to_taxonomy.pl is
   required input to from_vecscreen_to_summary.pl.

8. compare_vector_matches_wtaxa.pl
   Program to classify vecscreen matches that have already been parsed
   with from_vecscreen_to_summary.pl. Matches can be classified as:

 A. NO_DATA:           there is no data about the source of the vector
                       segment in the match.

 B. TRUE_ARTIFICIAL:   the vector segment matched is an ARTIFICIAL
                       sequence and hence the match is TRUE
                       contamination.

 C. TRUE_ARTIFICIAL_MICROSAT: the vector segment matched is an
                              ARTIFICIAL sequence and hence the match
                              is TRUE contamination and the vector
                              contains a microsatellite.

 D. FALSE_AMR:         the query sequence is bacterial; the vector
                       segment matches a known sequence that confers
                       anti-microbial resistance and these can often
                       be transferred horizontally between bacteria
                       that may be taxonomically distant.

 E. FALSE_BIOLOGICAL:  the subject's biological origin is known and
                       its taxid is deemed close enough to that of the
                       query, so that the match is not contamination.

 F. TRUE_BIOLOGICAL :  the query and the matching subject originate from
                       taxa that are too far apart for the vector to occur
                       plausibly in the query.


RELEASE-NOTES.txt
# vecscreen_plus_taxonomy release notes 

-------------
0.17 release, minor update
December 21, 2020

- updates NCBI taxonomy tree (info-files/taxonomy_tree_wlevels.txt)
- fixes some typos in README
- adds this RELEASE-NOTES.txt file
-------------


scripts/test-scripts.pl
# test-scripts.pl: Testing script that runs example commands and compares output files 
#                  with expected files to make sure they are identical.
#
# EPN, Mon Jan 30 09:39:20 2017
# 
use strict;

# make *STDOUT file handle 'hot' so it automatically flushes whenever we print to it
select *STDOUT;
$| = 1;

# make sure the VECPLUSDIR environment variable is set
my $vecplusdir = $ENV{'VECPLUSDIR'};
if(! exists($ENV{'VECPLUSDIR'})) { 
  printf STDERR ("\nERROR, the environment variable VECPLUSDIR is not set, please set it to the directory that was created when you unpackaged the vecscreen_plus_taxonomy package.\n"); 
  exit(1); 
}
if(! (-d $vecplusdir)) { 
  printf STDERR ("\nERROR, the vecscreen_plus_taxonomy directory specified by your environment variable VECPLUSDIR does not exist.\n"); 
  exit(1); 
}    
my $vecplus_exec_dir = $vecplusdir . "/scripts/"; 

# make sure the executable files we need exist in $VECPLUSDIR
my %execs_H = (); # hash with paths to all required executables
$execs_H{"vecscreen"}                 = $vecplus_exec_dir . "vecscreen";
$execs_H{"srcchk"}                    = $vecplus_exec_dir . "srcchk";
$execs_H{"parse_vecscreen"}           = $vecplus_exec_dir . "parse_vecscreen.pl";
$execs_H{"combine_summaries"}         = $vecplus_exec_dir . "combine_summaries.pl";
$execs_H{"add_taxonomy"}              = $vecplus_exec_dir . "add_taxonomy.pl";
$execs_H{"from_vecscreen_to_summary"} = $vecplus_exec_dir . "from_vecscreen_to_summary.pl";
validate_executable_hash(\%execs_H);

########################################################################
# Ensure we have all input files and expected output files (to compare
# to actual output files we create) that we need.
########################################################################
# This needs to be manually updated. The code does not ensure that all 
# input files used in the 'Test *.pl' sections below are checked for
# in this block. So, if you add an input or expected file to a 'Test *.pl'
# block, add it here too.

my $command_width = 40;
printf("%-*s ... ", $command_width, "Checking that required input files exist");
my $input_dir             = "$VECPLUSDIR/test-files";
my $input_root            = "test";
my $input_dir_and_root    = $input_dir . "/" . $input_root;
my @input_suffixes_A      = ("input_sequence_file.fa");
my $input_dir2            = "$VECPLUSDIR/info-files";
my $input_dir_and_root2   = $input_dir . "/" . $input_root;
my @input_suffixes2_A     = ("taxonomy_tree_wlevels.txt"); # this file is too big to require a copy to exist in test-files

my $expected_dir          = "$VECPLUSDIR/test-files";
my $expected_root         = "expected";
my $expected_dir_and_root = $expected_dir . "/" . $expected_root;
my @expected_suffixes_A   = ("output_combined.txt", "output_combined.verbose.txt", "add_taxonomy.out", "add_taxonomy.verbose.out");

my $output_dir          = "$VECPLUSDIR/test-files";
my $output_root         = "tmp";
my $output_dir_and_root = $output_dir . "/" . $output_root;

my $output_dir2          = "./";
my $output_root2         = "tmp";
my $output_dir_and_root2 = $output_dir2 . "/" . $output_root2;

my $errmsg = "";

# check input files exist
foreach my $suffix (@input_suffixes_A) { 
  $errmsg .= check_file_exists_and_is_nonempty($input_dir . "/" . $input_root . "." . $suffix, "input");
}
foreach my $suffix (@input_suffixes2_A) { 
  $errmsg .= check_file_exists_and_is_nonempty($input_dir2 . "/" . $suffix, "input");
}
foreach my $file (@input_other_files_A) { 
  $errmsg .= check_file_exists_and_is_nonempty($file, "input");
}
# check expected output files exist
foreach my $suffix (@expected_suffixes_A) { 
  $errmsg .= check_file_exists_and_is_nonempty($expected_dir . "/" . $expected_root . "." . $suffix, "expected output");
}
# die if any were missing
if($errmsg ne "") { 
  die "\n" . $errmsg;
}
printf("done.\n");

###########################
# Test combine_summaries.pl
###########################
printf("%-*s ... ", $command_width, "Testing combine_summaries.pl");
######################
my $cmd = $execs_H{"combine_summaries"} . " --input_terminal $input_dir_and_root.output_internal.txt --input_internal $input_dir_and_root.output_terminal.txt --outfile $output_dir_and_root.output_combined.txt > /dev/null"; # standard output is expected to be empty
run_command($cmd, 0);
my @out_A = ("$output_dir_and_root.output_combined.txt");
check_many_files_exist_and_are_nonempty(\@out_A, "output");
diff_files("$output_dir_and_root.output_combined.txt", "$expected_dir_and_root.output_combined.txt");
rm_files(\@out_A);

#####################
# same command, but with --verbose added
my $cmd = $execs_H{"combine_summaries"} . " --input_terminal $input_dir_and_root.output_internal.txt --input_internal $input_dir_and_root.output_terminal.txt --outfile $output_dir_and_root.output_combined.verbose.txt --verbose > /dev/null"; # standard output is expected to be empty
run_command($cmd, 0);
@out_A = ("$output_dir_and_root.output_combined.verbose.txt");
check_many_files_exist_and_are_nonempty(\@out_A, "output");
diff_files("$output_dir_and_root.output_combined.verbose.txt", "$expected_dir_and_root.output_combined.verbose.txt");
rm_files(\@out_A);

#####################
printf("done.\n");

######################
# Test add_taxonomy.pl
######################
printf("%-*s ... ", $command_width, "Testing add_taxonomy.pl");
######################
my $cmd = $execs_H{"add_taxonomy"} . " --input_summary $input_dir_and_root.combine_summaries --input_taxa taxonomy_tree_wlevels.txt --outfile $output_dir_and_root.add_taxonomy.out"; 
run_command($cmd, 0);
@out_A = ("$output_dir_and_root.add_taxonomy.out");
check_many_files_exist_and_are_nonempty(\@out_A, "output");
diff_files("$output_dir_and_root.add_taxonomy.out", "$expected_dir_and_root.add_taxonomy.out");
rm_files(\@out_A);

#####################
# same command, but with --verbose added
my $cmd = $execs_H{"add_taxonomy"} . " --input_summary $input_dir_and_root.combine_summaries --input_taxa taxonomy_tree_wlevels.txt --outfile $output_dir_and_root.add_taxonomy.verbose.out --verbose"; 
run_command($cmd, 0);
@out_A = ("$output_dir_and_root.add_taxonomy.verbose.out");
check_many_files_exist_and_are_nonempty(\@out_A, "output");
diff_files("$output_dir_and_root.add_taxonomy.verbose.out", "$expected_dir_and_root.add_taxonomy.verbose.out");
rm_files(\@out_A);

#####################
printf("done.\n");

###################################
# Test from_vecscreen_to_summary.pl
###################################
printf("%-*s ... ", $command_width, "Testing from_vecscreen_to_summary.pl");
######################
my $cmd = $execs_H{"from_vecscreen_to_summary"} . " --keep --output_root tmp --input_fasta $input_dir_and_root.input_sequence_file.fa --input_taxa taxonomy_tree_wlevels.txt --verbose --combine_output > /dev/null"; # standard output is expected to be empty
run_command($cmd, 0);
# TODO: - make from_vecscreen_to_summary.pl create and output to a directory
#       - have from_vecscreen_to_summary.pl make sure all necessary files for all steps exist before starting step 1
#       - update this file to test from_vecscreen_to_summary.pl
#       - add parse_vecscreen.pl tests to this file
my @out_A = ("$output_dir_and_root2.output_combined_wtaxonomy.txt");
check_many_files_exist_and_are_nonempty(\@out_A, "output");
diff_files("$output_dir_and_root2.output_combined_wtaxonomy.txt", "$expected_dir_and_root.output_combined_wtaxonomy.txt");
rm_files(\@out_A);

printf("# All tests passed.\n");
printf("# SUCCESS\n");

#################################################################
# Subroutine:  check_many_files_exists_and_are_nonempty()
# Incept:      EPN, Mon Jan 30 10:26:28 2017
#
# Purpose:     Checks if several expected output files exist and 
#              are non-empty.
#
# Arguments:
#   $file_AR:  ref to array of files to check for
#   $type:     type of file (e.g. 'input' or 'output')
#
# Returns:    nothing
#
# Dies:       If >= 1 of the files does not exist. 
#################################################################
sub check_many_files_exist_and_are_nonempty { 
  my $sub_name = "check_many_files_exists_and_are_nonempty()";
  my $nargs_expected = 2;
  if(scalar(@_) != $nargs_expected) { printf STDERR ("ERROR, $sub_name entered with %d != %d input arguments.\n", scalar(@_), $nargs_expected); exit(1); } 

  my ($file_AR, $type) = @_;

  foreach my $file (@{$file_AR}) { 
    $errmsg .= check_file_exists_and_is_nonempty($file, $type);
  }
  if($errmsg ne "") { 
    die "\n" . $errmsg;
  }

  return;
}

#################################################################
# Subroutine:  check_file_exists_and_is_nonempty()
# Incept:      EPN, Mon Jan 30 09:15:06 2017
#
# Purpose:     Checks if an expected output file exists and 
#              is non-empty.
#
# Arguments:
#   $file:     file to check
#   $type:     type of file (e.g. 'input' or 'output')
#
# Returns:    error message if file does not exist or is empty,
#             if "", then file exists and is non-empty
#
# Dies:       Never
#################################################################
sub check_file_exists_and_is_nonempty { 
  my $sub_name = "check_file_exists_and_is_nonempty()";
  my $nargs_expected = 2;
  if(scalar(@_) != $nargs_expected) { printf STDERR ("ERROR, $sub_name entered with %d != %d input arguments.\n", scalar(@_), $nargs_expected); exit(1); } 

  my ($file, $type) = @_;

  if   (! -e $file) { return "ERROR, $type file $file does not exist\n"; }
  elsif(! -s $file) { return "ERROR, $type file $file exists but is empty\n"; }

  return "";
}

#################################################################
# Subroutine:  diff_files()
# Incept:      EPN, Mon Jan 30 09:03:43 2017
#
# Purpose:     Diffs two files and reports on any differences.
#
# Arguments:
#   $file1:    first  file to compare
#   $file2:    second file to compare
#
# Returns:    output from diff, if "", then both files are equal
#
# Dies:       if either $file1 or $file2 does not exist
#################################################################
sub diff_files { 
  my $sub_name = "diff_files()";
  my $nargs_expected = 2;
  if(scalar(@_) != $nargs_expected) { printf STDERR ("ERROR, $sub_name entered with %d != %d input arguments.\n", scalar(@_), $nargs_expected); exit(1); } 

  my ($file1, $file2) = @_;

  if(! -e $file1) { die "ERROR, file $file1 in $sub_name does not exist"; }
  if(! -e $file2) { die "ERROR, file $file2 in $sub_name does not exist"; }

  return `diff $file1 $file2`;
}

#################################################################
# Subroutine:  run_command()
# Incept:      EPN, Mon Dec 19 10:43:45 2016
#
# Purpose:     Runs a command using system() and exits in error 
#              if the command fails. 
#
# Arguments:
#   $cmd:         command to run, with a "system" command;
#   $echo_cmd:    '1' to echo command to screen, '0' not to
#
# Returns:    amount of time the command took, in seconds
#
# Dies:       if $cmd fails
#################################################################
sub run_command {
  my $sub_name = "run_command()";
  my $nargs_expected = 2;
  if(scalar(@_) != $nargs_expected) { printf STDERR ("ERROR, $sub_name entered with %d != %d input arguments.\n", scalar(@_), $nargs_expected); exit(1); } 

  my ($cmd, $echo_cmd) = @_;

  if($echo_cmd) { 
    printf("\tExecuting command: $cmd\n");
  }
  system($cmd);

  if($? != 0) { 
    die "ERROR in $sub_name, the following command failed:\n$cmd\n";
  }

  return;
}

#################################################################
# Subroutine:  rm_files
# Incept:      EPN, Mon Jan 30 11:10:58 2017
#
# Purpose:     Remove the files listed in @{$file_AR}
#              using unlink.
#
# Arguments:
#   $file_AR:  ref to array of files to check for
#
# Returns:    nothing
#
# Dies:       Never.
#################################################################
sub rm_files { 
  my $sub_name = "rm_files()";
  my $nargs_expected = 1;
  if(scalar(@_) != $nargs_expected) { printf STDERR ("ERROR, $sub_name entered with %d != %d input arguments.\n", scalar(@_), $nargs_expected); exit(1); } 

  my ($file_AR) = @_;

  foreach my $file (@{$file_AR}) { 
    unlink $file; 
  }

  return;
}

#################################################################
# Subroutine : validate_executable_hash()
#
# Purpose:     Given a reference to a hash in which the 
#              values are paths to executables, validate
#              that those files are executable.
#
# Arguments: 
#   $execs_HR: REF to hash, keys are short names to executable
#              e.g. "vecscreen", values are full paths to that
#              executable, e.g. "/usr/local/infernal/1.1.1/bin/vecscreen"
# 
# Returns:     void
#
# Dies:        if one or more executables does not exist
#
################################################################# 
sub validate_executable_hash { 
  my $nargs_expected = 1;
  my $sub_name = "validate_executable_hash()";
  if(scalar(@_) != $nargs_expected) { printf STDERR ("ERROR, $sub_name entered with %d != %d input arguments.\n", scalar(@_), $nargs_expected); exit(1); } 
  my ($execs_HR) = (@_);

  my $fail_str = undef;
  foreach my $key (sort keys %{$execs_HR}) { 
    if(! -e $execs_HR->{$key}) { 
      $fail_str .= "\t$execs_HR->{$key} does not exist.\n"; 
    }
    elsif(! -x $execs_HR->{$key}) { 
      $fail_str .= "\t$execs_HR->{$key} exists but is not an executable file.\n"; 
    }
  }
  
  if(defined $fail_str) { 
    die "ERROR in $sub_name(),\n$fail_str"; 
  }

  return;
}
